lib/activerecord/delay_touching.rb

```diff
@@ -61,6 +61,9 @@ def self.call
     # Apply the touches that were delayed.
     def self.apply
       begin
+        # If we don't do this then an infinite loop is possible due to how Set#subtract and ActiveRecord::Core#== work
```

Comment:
Not to nit pick, but I think Set#subtract and ActiveRecord::Core#== work as expected (even together), it's just that the object unintuitively changes in memory after it was used as a hash key. The fact that Hash#rehash already exists means that it's not that uncommon an issue; the problem here is that effects of the id-unsetting logic in ActiveRecord are not immediately obvious.

Would be in favor of clarifying this comment (or clarifying it and moving it to the definition of `remove_unpersisted_records`).

Reply:
Great point.